packages/jubmoji-api/src/proofs.ts

```diff
@@ -0,0 +1,313 @@
+    N,
+    pathToCircuits,
+  }: NUniqueJubmojiInCollectionClassArgs) {
+    this.collectionPubKeys = collectionPubKeys;
+    this.sigNullifierRandomness = sigNullifierRandomness;
+    this.N = N;
+    this.pathToCircuits = pathToCircuits;
+  }
+
+  async prove({
+    jubmojis,
+  }: NUniqueJubmojiInCollectionProofArgs): Promise<NUniqueJubmojiInCollectionProof> {
+    const pubKeyNullifierRandomness = getRandomNullifierRandomness();
+
+    const sigs = [];
+    const msgHashes = [];
+    const publicInputs = [];
+    const indices = [];
+    for (const jubmoji of jubmojis) {
+      const { sig, msgHash, pubKey, R, T, U } =
+        getMembershipProofArgsFromJubmoji(jubmoji);
+      sigs.push(sig);
+      msgHashes.push(msgHash);
+      publicInputs.push({ R, T, U });
+      indices.push(this.collectionPubKeys.indexOf(pubKey));
+    }
+
+    const merkleProofs = getMerkleProofListFromCache(
+      this.collectionPubKeys,
+      indices
+    );
+
+    const membershipProofs = await batchProveMembership({
+      sigs,
+      msgHashes,
+      publicInputs,
+      merkleProofs,
+      sigNullifierRandomness: hexToBigInt(this.sigNullifierRandomness),
+      pubKeyNullifierRandomness,
+      pathToCircuits: this.pathToCircuits,
+    });
+
+    return {
+      serializedMembershipProofs: membershipProofs.map(
+        serializeMembershipProof
+      ),
+    };
+  }
+
+  // Checks that all the membershipProofs are valid, and that their pubKeyNullifiers are unique
+  // while having the same pubKeyNullifierRandomnessHash
+  async verify({
+    serializedMembershipProofs,
+    usedSigNullifiers,
+  }: NUniqueJubmojiInCollectionProof): Promise<VerificationResult> {
+    // Must have at least N proofs
+    if (serializedMembershipProofs.length < this.N) {
+      return { verified: false };
+    }
+
+    const membershipProofs = serializedMembershipProofs.map(
+      deserializeMembershipProof
+    );
+
+    // Checks that all the pubKeyNullifiers are unique but all the
+    // pubKeyNullifierRandomnessHashes are the same
+    const pubKeyNullifiers = membershipProofs.map(
+      (proof) => getPublicSignalsFromMembershipZKP(proof.zkp).pubKeyNullifier
+    );
+    if (!areAllBigIntsDifferent(pubKeyNullifiers)) {
+      return { verified: false };
+    }
+
+    const pubKeyNullifierRandomnessHashes = membershipProofs.map(
+      (proof) =>
+        getPublicSignalsFromMembershipZKP(proof.zkp)
+          .pubKeyNullifierRandomnessHash
+    );
+    if (!areAllBigIntsTheSame(pubKeyNullifierRandomnessHashes)) {
+      return { verified: false };
+    }
+
+    const merkleRoot = getMerkleRootFromCache(this.collectionPubKeys);
+
+    return await batchVerifyMembership({
+      proofs: membershipProofs,
+      merkleRoot,
+      sigNullifierRandomness: hexToBigInt(this.sigNullifierRandomness),
+      usedSigNullifiers: usedSigNullifiers?.map(hexToBigInt),
+      pathToCircuits: this.pathToCircuits,
+    });
+  }
+}
+
+export class PublicMessageSignature
```

Comment:
We should add an issue to test this with the cards we get. We might actually only have arbitrary signing access via the secp256k1 keys on the card (I forgot) so unclear what this implementation will look like rn.

Reply:
https://github.com/project-vinyl/jubmoji.quest/issues/35